Scheduler crash

In my last test run, scheduler crashed with the following error:
```
E0718 07:43:44.372428    3537 runtime.go:70] Observed a panic: "Key found in assumed set but not in podStates. Potentially a logical error." (Key found in assumed set but not in podStates. Potentially a logical error.)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:64
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:48
/usr/local/go/src/runtime/asm_amd64.s:472
/usr/local/go/src/runtime/panic.go:443
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache/cache.go:321
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache/cache.go:309
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache/cache.go:305
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:86
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:87
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:49
/usr/local/go/src/runtime/asm_amd64.s:1998
panic: Key found in assumed set but not in podStates. Potentially a logical error. [recovered]
        panic: Key found in assumed set but not in podStates. Potentially a logical error.

goroutine 57 [running]:
panic(0x169fb40, 0xc82ccccd00)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
k8s.io/kubernetes/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:55 +0x153
panic(0x169fb40, 0xc82ccccd00)
        /usr/local/go/src/runtime/panic.go:443 +0x4e9
k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache.(*schedulerCache).cleanupAssumedPods(0xc820015640, 0xecf1e7fb0, 0x16314887, 0x29b7380)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache/cache.go:321 +0x1a7
k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache.(*schedulerCache).cleanupExpiredAssumedPods(0xc820015640)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache/cache.go:309 +0x4f
k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache.(*schedulerCache).(k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache.cleanupExpiredAssumedPods)-fm()
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache/cache.go:305 +0x20
k8s.io/kubernetes/pkg/util/wait.JitterUntil.func1(0xc820541bb0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:86 +0x4b
k8s.io/kubernetes/pkg/util/wait.JitterUntil(0xc820541bb0, 0x3b9aca00, 0x0, 0x1, 0xc820495e00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:87 +0xb4
k8s.io/kubernetes/pkg/util/wait.Until(0xc820541bb0, 0x3b9aca00, 0xc820495e00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:49 +0x43
created by k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache.(*schedulerCache).run
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache/cache.go:305 +0x7b
```

I think my previous fix in #28886 wasn't fully correct. I will try to fix it later today.

@xiang90 @hongchaodeng @davidopp

--- a/plugin/pkg/scheduler/schedulercache/cache.go
+++ b/plugin/pkg/scheduler/schedulercache/cache.go
@@ -149,6 +149,7 @@ func (cache *schedulerCache) ForgetPod(pod *api.Pod) error {
 		if err != nil {
 			return err
 		}
+		delete(cache.assumedPods, key)
 		delete(cache.podStates, key)
 	default:
 		return fmt.Errorf("pod state wasn't assumed but get forgotten. Pod key: %v", key)

--- a/plugin/pkg/scheduler/schedulercache/cache_test.go
+++ b/plugin/pkg/scheduler/schedulercache/cache_test.go
@@ -457,6 +457,36 @@ func TestRemovePod(t *testing.T) {
 	}
 }
 
+func TestForgetPod(t *testing.T) {
+	nodeName := "node"
+	basePod := makeBasePod(nodeName, "test", "100m", "500", []api.ContainerPort{{HostPort: 80}})
+	tests := []struct {
+		pods []*api.Pod
+	}{{
+		pods: []*api.Pod{basePod},
+	}}
+	now := time.Now()
+	ttl := 10 * time.Second
+
+	for i, tt := range tests {
+		cache := newSchedulerCache(ttl, time.Second, nil)
+		for _, pod := range tt.pods {
+			if err := cache.assumePod(pod, now); err != nil {
+				t.Fatalf("assumePod failed: %v", err)
+			}
+		}
+		for _, pod := range tt.pods {
+			if err := cache.ForgetPod(pod); err != nil {
+				t.Fatalf("ForgetPod failed: %v", err)
+			}
+		}
+		cache.cleanupAssumedPods(now.Add(2 * ttl))
+		if n := cache.nodes[nodeName]; n != nil {
+			t.Errorf("#%d: expecting pod deleted and nil node info, get=%s", i, n)
+		}
+	}
+}
+
 func BenchmarkList1kNodes30kPods(b *testing.B) {
 	cache := setupCacheOf1kNodes30kPods(b)
 	b.ResetTimer()

